pkg/kubelet/kubelet_pods.go

```diff
@@ -84,20 +86,34 @@ func (kl *Kubelet) getActivePods() []*v1.Pod {
 }
 
 // makeDevices determines the devices for the given container.
-// Experimental. For now, we hardcode /dev/nvidia0 no matter what the user asks for
-// (we only support one device per node).
-// TODO: add support for more than 1 GPU after #28216.
-func makeDevices(container *v1.Container) []kubecontainer.DeviceInfo {
-	nvidiaGPULimit := container.Resources.Limits.NvidiaGPU()
-	if nvidiaGPULimit.Value() != 0 {
-		return []kubecontainer.DeviceInfo{
-			{PathOnHost: "/dev/nvidia0", PathInContainer: "/dev/nvidia0", Permissions: "mrw"},
-			{PathOnHost: "/dev/nvidiactl", PathInContainer: "/dev/nvidiactl", Permissions: "mrw"},
-			{PathOnHost: "/dev/nvidia-uvm", PathInContainer: "/dev/nvidia-uvm", Permissions: "mrw"},
-		}
+// Experimental.
+func (kl *Kubelet) makeDevices(pod *v1.Pod, container *v1.Container) ([]kubecontainer.DeviceInfo, error) {
+	if container.Resources.Limits.NvidiaGPU().IsZero() {
+		return nil, nil
 	}
 
-	return nil
+	nvidiaGPUPaths, err := kl.gpuManager.AllocateGPU(pod, container)
+	if err != nil {
+		return nil, err
+	}
+	devices := []kubecontainer.DeviceInfo{
+		{
+			PathOnHost:      nvidia.NvidiaCtlDevice,
+			PathInContainer: nvidia.NvidiaCtlDevice,
+			Permissions:     "mrw",
+		},
+		{
+			PathOnHost:      nvidia.NvidiaUVMDevice,
+			PathInContainer: nvidia.NvidiaUVMDevice,
+			Permissions:     "mrw",
+		},
+	}
+
+	for i, path := range nvidiaGPUPaths {
+		devices = append(devices, kubecontainer.DeviceInfo{PathOnHost: path, PathInContainer: "/dev/nvidia" + strconv.Itoa(i), Permissions: "mrw"})
+	}
+
```

Comment:
I hope there are valid reasons for why the CUDA driver needs one to one mapping between `major:minor` and character device files. It's an odd requirement from Linux perspective though.

> This is a little bit tricky and that's why generally you don't want to deal with the GPU runtime inside Kubernetes.

This is unavoidable. Kubernetes will eventually have to understand hardware topology to provide the best possible performance latency and isolation for pods. It is not possible to make GPUs as an isolated extensible resource and have it be managed independent of other resources. 
There are other means to extend GPU management in kubernetes which might help to hide some of the complexity associated with managing Nvidia GPUs.

Thanks for the tip on `nvidia-uvm`. I will update the PR to add it if present letting the overall node configuration unvalidated for now.